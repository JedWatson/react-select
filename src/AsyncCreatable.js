import PropTypes from 'prop-types';
import React from 'react';

import Async from './Async';
import Creatable from './Creatable';
import Select from './Select';

class AsyncCreatableSelect extends React.Component {

	focus () {
		this.select.focus();
	}

	render () {
		return (
			<Async {...this.props}>
				{({ ref, ...asyncProps }) => {
					const asyncRef = ref;
					return (<Creatable {...asyncProps} >
						{({ ref, ...creatableProps }) => {
							const creatableRef = ref;
							return this.props.children({
								...creatableProps,
								ref: (select) => {
									creatableRef(select);
									asyncRef(select);
									this.select = select;
								}
							});
						}}
					</Creatable>);
				}}
			</Async>
		);
	}
}

const defaultChildren = props => <Select {...props} />;

AsyncCreatableSelect.propTypes = {
	children: PropTypes.func.isRequired, // Child function responsible for creating the inner Select component; (props: Object): PropTypes.element
};

AsyncCreatableSelect.defaultProps = {
	children: defaultChildren,
};

export default AsyncCreatableSelect;
