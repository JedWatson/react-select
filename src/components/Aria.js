// @flow
import React from 'react';

import { SROnly } from '../primitives';
import { className } from '../utils';

type StatusProps = { availableResults: number };
export const AriaStatus = ({ availableResults }: StatusProps) => (
  <SROnly
    aria-atomic="true"
    aria-live="polite"
    role="status"
    className={className('sr-only')}
  >
    {availableResults} results are available.
  </SROnly>
);
