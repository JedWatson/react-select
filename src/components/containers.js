// @flow
import React, { Component, type Node, type ElementRef } from 'react';

import { Div } from '../primitives';
import { spacing } from '../theme';
import type { CommonProps, KeyboardEventHandler } from '../types';

// ==============================
// Root Container
// ==============================

type ContainerState = {
  /** Whether the select is disabled. */
  isDisabled: boolean,
  /** Whether the text in the select is indented from right to left. */
  isRtl: boolean,
};

export type ContainerProps = CommonProps &
  ContainerState & {
    /** The children to be rendered. */
    children: Node,
    /** Inner props to be passed down to the container. */
    innerProps: { onKeyDown: KeyboardEventHandler },
  };
export const containerCSS = ({ isDisabled, isRtl }: ContainerState) => ({
  direction: isRtl ? 'rtl' : null,
  pointerEvents: isDisabled ? 'none' : null, // cancel mouse events when disabled
  position: 'relative',
});
export const SelectContainer = (props: ContainerProps) => {
  const { children, cx, getStyles, innerProps, isDisabled, isRtl } = props;
  return (
    <Div
      css={getStyles('container', props)}
      className={cx('', { isDisabled, isRtl })}
      {...innerProps}
    >
      {children}
    </Div>
  );
};

// ==============================
// Value Container
// ==============================

export type ValueContainerProps = CommonProps & {
  /** Set when the value container should hold multiple values. This is important for styling. */
  isMulti: boolean,
  /** Whether the value container currently holds a value. */
  hasValue: boolean,
  /** Whether there should be a maximum height to the container */
  maxHeight: number,
  /** The children to be rendered. */
  children: Node,
};
export const valueContainerCSS = ({ maxHeight }: ValueContainerProps) => ({
  alignItems: 'center',
  display: 'flex',
  flex: 1,
  flexWrap: 'wrap',
  maxHeight: maxHeight, // max-height allows scroll when multi
  overflowY: 'auto',
  padding: `${spacing.baseUnit / 2}px ${spacing.baseUnit * 2}px`,
  WebkitOverflowScrolling: 'touch',
  position: 'relative',
});
export class ValueContainer extends Component<ValueContainerProps> {
  shouldScrollBottom: boolean = false;
  node: HTMLElement;
  componentWillUpdate() {
    if (!this.props.isMulti) return;

    // scroll only if the user was already at the bottom
    const total = this.node.scrollTop + this.node.offsetHeight;
    this.shouldScrollBottom = total === this.node.scrollHeight;
  }
  componentDidUpdate() {
    if (!this.props.isMulti) return;

    // ensure we're showing items being added by forcing scroll to the bottom
    if (this.shouldScrollBottom && this.node) {
      this.node.scrollTop = this.node.scrollHeight;
    }
  }
  getScrollContainer = (ref: ElementRef<*>) => {
    this.node = ref;
  };
  render() {
    const { children, cx, isMulti, getStyles, hasValue } = this.props;

    return (
      <Div
        innerRef={isMulti ? this.getScrollContainer : undefined}
        className={cx('value-container', { isMulti, hasValue })}
        css={getStyles('valueContainer', this.props)}
      >
        {children}
      </Div>
    );
  }
}

// ==============================
// Indicator Container
// ==============================

type IndicatorsState = {
  /** Whether the text should be rendered right to left. */
  isRtl: boolean,
};

export type IndicatorContainerProps = CommonProps &
  IndicatorsState & {
    /** The children to be rendered. */
    children: Node,
  };

export const indicatorsContainerCSS = () => ({
  alignItems: 'center',
  alignSelf: 'stretch',
  display: 'flex',
  flexShrink: 0,
});
export const IndicatorsContainer = (props: IndicatorContainerProps) => {
  const { children, cx, getStyles } = props;

  return (
    <Div
      className={cx('indicators')}
      css={getStyles('indicatorsContainer', props)}
    >
      {children}
    </Div>
  );
};
