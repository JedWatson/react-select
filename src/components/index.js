// @flow
import {
  IndicatorsContainer,
  SelectContainer,
  ValueContainer,
} from './containers';
import { ClearIndicator, DropdownIndicator } from './indicators';

import Control from './Control';
import Group from './Group';
import Input from './Input';
import Label from './Label';
import Menu, { MenuList, NoOptions } from './Menu';
import MultiValue from './MultiValue';
import Option from './Option';
import Placeholder from './Placeholder';
import SingleValue from './SingleValue';

export type SelectComponents = {
  ClearIndicator: typeof ClearIndicator,
  Control: typeof Control,
  DropdownIndicator: typeof DropdownIndicator,
  Group: typeof Group,
  Label: typeof Label,
  IndicatorsContainer: typeof IndicatorsContainer,
  Input: typeof Input,
  Menu: typeof Menu,
  MenuList: typeof MenuList,
  MultiValue: typeof MultiValue,
  NoOptions: typeof NoOptions,
  Option: typeof Option,
  Placeholder: typeof Placeholder,
  SelectContainer: typeof SelectContainer,
  SingleValue: typeof SingleValue,
  ValueContainer: typeof ValueContainer,
};

export const components: SelectComponents = {
  ClearIndicator: ClearIndicator,
  Control: Control,
  DropdownIndicator: DropdownIndicator,
  Group: Group,
  Label: Label,
  IndicatorsContainer: IndicatorsContainer,
  Input: Input,
  Menu: Menu,
  MenuList: MenuList,
  MultiValue: MultiValue,
  NoOptions: NoOptions,
  Option: Option,
  Placeholder: Placeholder,
  SelectContainer: SelectContainer,
  SingleValue: SingleValue,
  ValueContainer: ValueContainer,
};

type Props = {
  components: SelectComponents,
};

export const defaultComponents = (props: Props) => ({
  ...components,
  ...props.components,
});
