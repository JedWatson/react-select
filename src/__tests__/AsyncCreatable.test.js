import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import AsyncCreatable from '../AsyncCreatable';
import Select from '../Select';
import { components } from '../components';
import { OPTIONS } from './constants';
const { Menu, Option } = components;

test('defaults - snapshot', () => {
  const tree = mount(<AsyncCreatable />);
  expect(toJson(tree)).toMatchSnapshot();
});

test('creates an inner Select', () => {
  const asyncCreatableWrapper = mount(
    <AsyncCreatable className="react-select" classNamePrefix="react-select"/>
  );
  expect(asyncCreatableWrapper.find(Select).exists()).toBeTruthy();
});

test('render decorated select with props passed', () => {
  const asyncCreatableWrapper = mount(<AsyncCreatable className="foo" classNamePrefix="foo" />);
  expect(asyncCreatableWrapper.find(Select).props().className).toBe('foo');
});

test('to show the create option in menu', () => {
  let asyncCreatableWrapper = mount(
    <AsyncCreatable className="react-select" classNamePrefix="react-select"/>
  );
  let inputValueWrapper = asyncCreatableWrapper.find(
    'div.react-select__input input'
  );
  asyncCreatableWrapper.setProps({ inputValue: 'a' });
  inputValueWrapper.simulate('change', { currentTarget: { value: 'a' } });
  expect(
    asyncCreatableWrapper
      .find(Option)
      .last()
      .text()
  ).toBe('Create "a"');
});

test('to show loading and then create option in menu', () => {
  jest.useFakeTimers();
  let loadOptionsSpy = jest.fn((inputValue, callback) =>
    setTimeout(() => callback(OPTIONS), 200)
  );
  let asyncCreatableWrapper = mount(
    <AsyncCreatable className="react-select" classNamePrefix="react-select" loadOptions={loadOptionsSpy} />
  );
  let inputValueWrapper = asyncCreatableWrapper.find(
    'div.react-select__input input'
  );
  asyncCreatableWrapper.setProps({ inputValue: 'a' });
  inputValueWrapper.simulate('change', { currentTarget: { value: 'a' } });

  // to show a loading message while loading options
  expect(asyncCreatableWrapper.find(Menu).text()).toBe('Loading...');
  jest.runAllTimers();
  asyncCreatableWrapper.update();

  // show create options once options are loaded
  expect(
    asyncCreatableWrapper
      .find(Option)
      .last()
      .text()
  ).toBe('Create "a"');
});
