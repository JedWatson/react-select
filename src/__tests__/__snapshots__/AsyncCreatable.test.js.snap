// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults - snapshot 1`] = `
<Async
  cacheOptions={false}
  defaultOptions={false}
>
  <StateManager
    cacheOptions={false}
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultOptions={false}
    defaultValue={null}
    filterOption={null}
    isLoading={false}
    onInputChange={[Function]}
    options={Array []}
  >
    <Creatable
      allowCreateWhileLoading={false}
      cacheOptions={false}
      defaultInputValue=""
      defaultMenuIsOpen={false}
      defaultOptions={false}
      defaultValue={null}
      filterOption={null}
      formatCreateLabel={[Function]}
      getNewOptionData={[Function]}
      inputValue=""
      isLoading={false}
      isValidNewOption={[Function]}
      menuIsOpen={false}
      onChange={[Function]}
      onInputChange={[Function]}
      onMenuClose={[Function]}
      onMenuOpen={[Function]}
      options={Array []}
      value={null}
    >
      <Select
        allowCreateWhileLoading={false}
        backspaceRemovesValue={true}
        blurInputOnSelect={true}
        cacheOptions={false}
        captureMenuScroll={false}
        closeMenuOnSelect={true}
        components={Object {}}
        defaultInputValue=""
        defaultMenuIsOpen={false}
        defaultOptions={false}
        defaultValue={null}
        escapeClearsValue={false}
        filterOption={null}
        formatCreateLabel={[Function]}
        formatGroupLabel={[Function]}
        getNewOptionData={[Function]}
        getOptionLabel={[Function]}
        getOptionValue={[Function]}
        hideSelectedOptions={true}
        inputValue=""
        isDisabled={false}
        isLoading={false}
        isMulti={false}
        isOptionDisabled={[Function]}
        isRtl={false}
        isSearchable={true}
        isValidNewOption={[Function]}
        loadingMessage={[Function]}
        maxMenuHeight={300}
        maxValueHeight={100}
        menuIsOpen={false}
        menuPlacement="bottom"
        menuPosition="absolute"
        menuShouldBlockScroll={false}
        menuShouldScrollIntoView={true}
        minMenuHeight={140}
        noOptionsMessage={[Function]}
        onChange={[Function]}
        onInputChange={[Function]}
        onMenuClose={[Function]}
        onMenuOpen={[Function]}
        openMenuOnClick={true}
        openMenuOnFocus={false}
        options={Array []}
        pageSize={5}
        placeholder="Select..."
        screenReaderStatus={[Function]}
        styles={Object {}}
        tabIndex="0"
        tabSelectsValue={true}
        value={null}
      >
        <SelectContainer
          clearValue={[Function]}
          cx={[Function]}
          getStyles={[Function]}
          getValue={[Function]}
          hasValue={false}
          innerProps={
            Object {
              "id": undefined,
              "onKeyDown": [Function],
            }
          }
          isDisabled={false}
          isFocused={false}
          isMulti={false}
          isRtl={false}
          options={Array []}
          selectOption={[Function]}
          selectProps={
            Object {
              "allowCreateWhileLoading": false,
              "backspaceRemovesValue": true,
              "blurInputOnSelect": true,
              "cacheOptions": false,
              "captureMenuScroll": false,
              "closeMenuOnSelect": true,
              "components": Object {},
              "defaultInputValue": "",
              "defaultMenuIsOpen": false,
              "defaultOptions": false,
              "defaultValue": null,
              "escapeClearsValue": false,
              "filterOption": null,
              "formatCreateLabel": [Function],
              "formatGroupLabel": [Function],
              "getNewOptionData": [Function],
              "getOptionLabel": [Function],
              "getOptionValue": [Function],
              "hideSelectedOptions": true,
              "inputValue": "",
              "isDisabled": false,
              "isLoading": false,
              "isMulti": false,
              "isOptionDisabled": [Function],
              "isRtl": false,
              "isSearchable": true,
              "isValidNewOption": [Function],
              "loadingMessage": [Function],
              "maxMenuHeight": 300,
              "maxValueHeight": 100,
              "menuIsOpen": false,
              "menuPlacement": "bottom",
              "menuPosition": "absolute",
              "menuShouldBlockScroll": false,
              "menuShouldScrollIntoView": true,
              "minMenuHeight": 140,
              "noOptionsMessage": [Function],
              "onChange": [Function],
              "onInputChange": [Function],
              "onMenuClose": [Function],
              "onMenuOpen": [Function],
              "openMenuOnClick": true,
              "openMenuOnFocus": false,
              "options": Array [],
              "pageSize": 5,
              "placeholder": "Select...",
              "screenReaderStatus": [Function],
              "styles": Object {},
              "tabIndex": "0",
              "tabSelectsValue": true,
              "value": null,
            }
          }
          setValue={[Function]}
        >
          <Component
            className="css-10nd86i "
            onKeyDown={[Function]}
          >
            <div
              className="css-10nd86i "
              onKeyDown={[Function]}
            >
              <A11yText
                aria-atomic="true"
                aria-live="polite"
                role="status"
              >
                <span
                  aria-atomic="true"
                  aria-live="polite"
                  className="css-1uvvha"
                  role="status"
                >
                  0 results available.
                </span>
              </A11yText>
              <Control
                clearValue={[Function]}
                cx={[Function]}
                getStyles={[Function]}
                getValue={[Function]}
                hasValue={false}
                innerProps={
                  Object {
                    "innerRef": [Function],
                    "onMouseDown": [Function],
                    "onTouchEnd": [Function],
                  }
                }
                isDisabled={false}
                isFocused={false}
                isMulti={false}
                isRtl={false}
                options={Array []}
                selectOption={[Function]}
                selectProps={
                  Object {
                    "allowCreateWhileLoading": false,
                    "backspaceRemovesValue": true,
                    "blurInputOnSelect": true,
                    "cacheOptions": false,
                    "captureMenuScroll": false,
                    "closeMenuOnSelect": true,
                    "components": Object {},
                    "defaultInputValue": "",
                    "defaultMenuIsOpen": false,
                    "defaultOptions": false,
                    "defaultValue": null,
                    "escapeClearsValue": false,
                    "filterOption": null,
                    "formatCreateLabel": [Function],
                    "formatGroupLabel": [Function],
                    "getNewOptionData": [Function],
                    "getOptionLabel": [Function],
                    "getOptionValue": [Function],
                    "hideSelectedOptions": true,
                    "inputValue": "",
                    "isDisabled": false,
                    "isLoading": false,
                    "isMulti": false,
                    "isOptionDisabled": [Function],
                    "isRtl": false,
                    "isSearchable": true,
                    "isValidNewOption": [Function],
                    "loadingMessage": [Function],
                    "maxMenuHeight": 300,
                    "maxValueHeight": 100,
                    "menuIsOpen": false,
                    "menuPlacement": "bottom",
                    "menuPosition": "absolute",
                    "menuShouldBlockScroll": false,
                    "menuShouldScrollIntoView": true,
                    "minMenuHeight": 140,
                    "noOptionsMessage": [Function],
                    "onChange": [Function],
                    "onInputChange": [Function],
                    "onMenuClose": [Function],
                    "onMenuOpen": [Function],
                    "openMenuOnClick": true,
                    "openMenuOnFocus": false,
                    "options": Array [],
                    "pageSize": 5,
                    "placeholder": "Select...",
                    "screenReaderStatus": [Function],
                    "styles": Object {},
                    "tabIndex": "0",
                    "tabSelectsValue": true,
                    "value": null,
                  }
                }
                setValue={[Function]}
              >
                <Component
                  className="css-1aya2g8 "
                  innerRef={[Function]}
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                >
                  <div
                    className="css-1aya2g8 "
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                  >
                    <ValueContainer
                      clearValue={[Function]}
                      cx={[Function]}
                      getStyles={[Function]}
                      getValue={[Function]}
                      hasValue={false}
                      isDisabled={false}
                      isMulti={false}
                      isRtl={false}
                      maxHeight={100}
                      options={Array []}
                      selectOption={[Function]}
                      selectProps={
                        Object {
                          "allowCreateWhileLoading": false,
                          "backspaceRemovesValue": true,
                          "blurInputOnSelect": true,
                          "cacheOptions": false,
                          "captureMenuScroll": false,
                          "closeMenuOnSelect": true,
                          "components": Object {},
                          "defaultInputValue": "",
                          "defaultMenuIsOpen": false,
                          "defaultOptions": false,
                          "defaultValue": null,
                          "escapeClearsValue": false,
                          "filterOption": null,
                          "formatCreateLabel": [Function],
                          "formatGroupLabel": [Function],
                          "getNewOptionData": [Function],
                          "getOptionLabel": [Function],
                          "getOptionValue": [Function],
                          "hideSelectedOptions": true,
                          "inputValue": "",
                          "isDisabled": false,
                          "isLoading": false,
                          "isMulti": false,
                          "isOptionDisabled": [Function],
                          "isRtl": false,
                          "isSearchable": true,
                          "isValidNewOption": [Function],
                          "loadingMessage": [Function],
                          "maxMenuHeight": 300,
                          "maxValueHeight": 100,
                          "menuIsOpen": false,
                          "menuPlacement": "bottom",
                          "menuPosition": "absolute",
                          "menuShouldBlockScroll": false,
                          "menuShouldScrollIntoView": true,
                          "minMenuHeight": 140,
                          "noOptionsMessage": [Function],
                          "onChange": [Function],
                          "onInputChange": [Function],
                          "onMenuClose": [Function],
                          "onMenuOpen": [Function],
                          "openMenuOnClick": true,
                          "openMenuOnFocus": false,
                          "options": Array [],
                          "pageSize": 5,
                          "placeholder": "Select...",
                          "screenReaderStatus": [Function],
                          "styles": Object {},
                          "tabIndex": "0",
                          "tabSelectsValue": true,
                          "value": null,
                        }
                      }
                      setValue={[Function]}
                    >
                      <Component
                        className="css-nbc25y "
                      >
                        <div
                          className="css-nbc25y "
                        >
                          <Placeholder
                            clearValue={[Function]}
                            cx={[Function]}
                            getStyles={[Function]}
                            getValue={[Function]}
                            hasValue={false}
                            isDisabled={false}
                            isMulti={false}
                            isRtl={false}
                            key="placeholder"
                            options={Array []}
                            selectOption={[Function]}
                            selectProps={
                              Object {
                                "allowCreateWhileLoading": false,
                                "backspaceRemovesValue": true,
                                "blurInputOnSelect": true,
                                "cacheOptions": false,
                                "captureMenuScroll": false,
                                "closeMenuOnSelect": true,
                                "components": Object {},
                                "defaultInputValue": "",
                                "defaultMenuIsOpen": false,
                                "defaultOptions": false,
                                "defaultValue": null,
                                "escapeClearsValue": false,
                                "filterOption": null,
                                "formatCreateLabel": [Function],
                                "formatGroupLabel": [Function],
                                "getNewOptionData": [Function],
                                "getOptionLabel": [Function],
                                "getOptionValue": [Function],
                                "hideSelectedOptions": true,
                                "inputValue": "",
                                "isDisabled": false,
                                "isLoading": false,
                                "isMulti": false,
                                "isOptionDisabled": [Function],
                                "isRtl": false,
                                "isSearchable": true,
                                "isValidNewOption": [Function],
                                "loadingMessage": [Function],
                                "maxMenuHeight": 300,
                                "maxValueHeight": 100,
                                "menuIsOpen": false,
                                "menuPlacement": "bottom",
                                "menuPosition": "absolute",
                                "menuShouldBlockScroll": false,
                                "menuShouldScrollIntoView": true,
                                "minMenuHeight": 140,
                                "noOptionsMessage": [Function],
                                "onChange": [Function],
                                "onInputChange": [Function],
                                "onMenuClose": [Function],
                                "onMenuOpen": [Function],
                                "openMenuOnClick": true,
                                "openMenuOnFocus": false,
                                "options": Array [],
                                "pageSize": 5,
                                "placeholder": "Select...",
                                "screenReaderStatus": [Function],
                                "styles": Object {},
                                "tabIndex": "0",
                                "tabSelectsValue": true,
                                "value": null,
                              }
                            }
                            setValue={[Function]}
                          >
                            <Component
                              className="css-1492t68 "
                            >
                              <div
                                className="css-1492t68 "
                              >
                                Select...
                              </div>
                            </Component>
                          </Placeholder>
                          <Input
                            aria-autocomplete="list"
                            aria-busy={false}
                            aria-expanded={false}
                            aria-haspopup={false}
                            autoCapitalize="none"
                            autoComplete="off"
                            autoCorrect="off"
                            cx={[Function]}
                            getStyles={[Function]}
                            id="react-select-2-input"
                            innerRef={[Function]}
                            isDisabled={false}
                            isHidden={false}
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            role="combobox"
                            spellCheck="false"
                            tabIndex="0"
                            type="text"
                            value=""
                          >
                            <Component
                              className="css-rsyb7x"
                            >
                              <div
                                className="css-rsyb7x"
                              >
                                <AutosizeInput
                                  aria-autocomplete="list"
                                  aria-busy={false}
                                  aria-expanded={false}
                                  aria-haspopup={false}
                                  autoCapitalize="none"
                                  autoComplete="off"
                                  autoCorrect="off"
                                  className=""
                                  disabled={false}
                                  id="react-select-2-input"
                                  injectStyles={true}
                                  inputRef={[Function]}
                                  inputStyle={
                                    Object {
                                      "background": 0,
                                      "border": 0,
                                      "color": "inherit",
                                      "fontSize": "inherit",
                                      "opacity": 1,
                                      "outline": 0,
                                      "padding": 0,
                                    }
                                  }
                                  minWidth={1}
                                  onBlur={[Function]}
                                  onChange={[Function]}
                                  onFocus={[Function]}
                                  role="combobox"
                                  spellCheck="false"
                                  tabIndex="0"
                                  type="text"
                                  value=""
                                >
                                  <div
                                    className=""
                                    style={
                                      Object {
                                        "display": "inline-block",
                                      }
                                    }
                                  >
                                    <input
                                      aria-autocomplete="list"
                                      aria-busy={false}
                                      aria-expanded={false}
                                      aria-haspopup={false}
                                      autoCapitalize="none"
                                      autoComplete="off"
                                      autoCorrect="off"
                                      disabled={false}
                                      id="react-select-2-input"
                                      onBlur={[Function]}
                                      onChange={[Function]}
                                      onFocus={[Function]}
                                      role="combobox"
                                      spellCheck="false"
                                      style={
                                        Object {
                                          "background": 0,
                                          "border": 0,
                                          "boxSizing": "content-box",
                                          "color": "inherit",
                                          "fontSize": "inherit",
                                          "opacity": 1,
                                          "outline": 0,
                                          "padding": 0,
                                          "width": "2px",
                                        }
                                      }
                                      tabIndex="0"
                                      type="text"
                                      value=""
                                    />
                                    <div
                                      style={
                                        Object {
                                          "height": 0,
                                          "left": 0,
                                          "overflow": "scroll",
                                          "position": "absolute",
                                          "top": 0,
                                          "visibility": "hidden",
                                          "whiteSpace": "pre",
                                        }
                                      }
                                    />
                                  </div>
                                </AutosizeInput>
                              </div>
                            </Component>
                          </Input>
                        </div>
                      </Component>
                    </ValueContainer>
                    <IndicatorsContainer
                      clearValue={[Function]}
                      cx={[Function]}
                      getStyles={[Function]}
                      getValue={[Function]}
                      hasValue={false}
                      isDisabled={false}
                      isMulti={false}
                      isRtl={false}
                      options={Array []}
                      selectOption={[Function]}
                      selectProps={
                        Object {
                          "allowCreateWhileLoading": false,
                          "backspaceRemovesValue": true,
                          "blurInputOnSelect": true,
                          "cacheOptions": false,
                          "captureMenuScroll": false,
                          "closeMenuOnSelect": true,
                          "components": Object {},
                          "defaultInputValue": "",
                          "defaultMenuIsOpen": false,
                          "defaultOptions": false,
                          "defaultValue": null,
                          "escapeClearsValue": false,
                          "filterOption": null,
                          "formatCreateLabel": [Function],
                          "formatGroupLabel": [Function],
                          "getNewOptionData": [Function],
                          "getOptionLabel": [Function],
                          "getOptionValue": [Function],
                          "hideSelectedOptions": true,
                          "inputValue": "",
                          "isDisabled": false,
                          "isLoading": false,
                          "isMulti": false,
                          "isOptionDisabled": [Function],
                          "isRtl": false,
                          "isSearchable": true,
                          "isValidNewOption": [Function],
                          "loadingMessage": [Function],
                          "maxMenuHeight": 300,
                          "maxValueHeight": 100,
                          "menuIsOpen": false,
                          "menuPlacement": "bottom",
                          "menuPosition": "absolute",
                          "menuShouldBlockScroll": false,
                          "menuShouldScrollIntoView": true,
                          "minMenuHeight": 140,
                          "noOptionsMessage": [Function],
                          "onChange": [Function],
                          "onInputChange": [Function],
                          "onMenuClose": [Function],
                          "onMenuOpen": [Function],
                          "openMenuOnClick": true,
                          "openMenuOnFocus": false,
                          "options": Array [],
                          "pageSize": 5,
                          "placeholder": "Select...",
                          "screenReaderStatus": [Function],
                          "styles": Object {},
                          "tabIndex": "0",
                          "tabSelectsValue": true,
                          "value": null,
                        }
                      }
                      setValue={[Function]}
                    >
                      <Component
                        className="css-1wy0on6 "
                      >
                        <div
                          className="css-1wy0on6 "
                        >
                          <IndicatorSeparator
                            clearValue={[Function]}
                            cx={[Function]}
                            getStyles={[Function]}
                            getValue={[Function]}
                            hasValue={false}
                            innerProps={
                              Object {
                                "role": "presentation",
                              }
                            }
                            isDisabled={false}
                            isFocused={false}
                            isMulti={false}
                            isRtl={false}
                            options={Array []}
                            selectOption={[Function]}
                            selectProps={
                              Object {
                                "allowCreateWhileLoading": false,
                                "backspaceRemovesValue": true,
                                "blurInputOnSelect": true,
                                "cacheOptions": false,
                                "captureMenuScroll": false,
                                "closeMenuOnSelect": true,
                                "components": Object {},
                                "defaultInputValue": "",
                                "defaultMenuIsOpen": false,
                                "defaultOptions": false,
                                "defaultValue": null,
                                "escapeClearsValue": false,
                                "filterOption": null,
                                "formatCreateLabel": [Function],
                                "formatGroupLabel": [Function],
                                "getNewOptionData": [Function],
                                "getOptionLabel": [Function],
                                "getOptionValue": [Function],
                                "hideSelectedOptions": true,
                                "inputValue": "",
                                "isDisabled": false,
                                "isLoading": false,
                                "isMulti": false,
                                "isOptionDisabled": [Function],
                                "isRtl": false,
                                "isSearchable": true,
                                "isValidNewOption": [Function],
                                "loadingMessage": [Function],
                                "maxMenuHeight": 300,
                                "maxValueHeight": 100,
                                "menuIsOpen": false,
                                "menuPlacement": "bottom",
                                "menuPosition": "absolute",
                                "menuShouldBlockScroll": false,
                                "menuShouldScrollIntoView": true,
                                "minMenuHeight": 140,
                                "noOptionsMessage": [Function],
                                "onChange": [Function],
                                "onInputChange": [Function],
                                "onMenuClose": [Function],
                                "onMenuOpen": [Function],
                                "openMenuOnClick": true,
                                "openMenuOnFocus": false,
                                "options": Array [],
                                "pageSize": 5,
                                "placeholder": "Select...",
                                "screenReaderStatus": [Function],
                                "styles": Object {},
                                "tabIndex": "0",
                                "tabSelectsValue": true,
                                "value": null,
                              }
                            }
                            setValue={[Function]}
                          >
                            <span
                              className="css-d8oujb "
                              role="presentation"
                            />
                          </IndicatorSeparator>
                          <DropdownIndicator
                            clearValue={[Function]}
                            cx={[Function]}
                            getStyles={[Function]}
                            getValue={[Function]}
                            hasValue={false}
                            innerProps={
                              Object {
                                "onMouseDown": [Function],
                                "onTouchEnd": [Function],
                                "role": "button",
                              }
                            }
                            isDisabled={false}
                            isFocused={false}
                            isMulti={false}
                            isRtl={false}
                            options={Array []}
                            selectOption={[Function]}
                            selectProps={
                              Object {
                                "allowCreateWhileLoading": false,
                                "backspaceRemovesValue": true,
                                "blurInputOnSelect": true,
                                "cacheOptions": false,
                                "captureMenuScroll": false,
                                "closeMenuOnSelect": true,
                                "components": Object {},
                                "defaultInputValue": "",
                                "defaultMenuIsOpen": false,
                                "defaultOptions": false,
                                "defaultValue": null,
                                "escapeClearsValue": false,
                                "filterOption": null,
                                "formatCreateLabel": [Function],
                                "formatGroupLabel": [Function],
                                "getNewOptionData": [Function],
                                "getOptionLabel": [Function],
                                "getOptionValue": [Function],
                                "hideSelectedOptions": true,
                                "inputValue": "",
                                "isDisabled": false,
                                "isLoading": false,
                                "isMulti": false,
                                "isOptionDisabled": [Function],
                                "isRtl": false,
                                "isSearchable": true,
                                "isValidNewOption": [Function],
                                "loadingMessage": [Function],
                                "maxMenuHeight": 300,
                                "maxValueHeight": 100,
                                "menuIsOpen": false,
                                "menuPlacement": "bottom",
                                "menuPosition": "absolute",
                                "menuShouldBlockScroll": false,
                                "menuShouldScrollIntoView": true,
                                "minMenuHeight": 140,
                                "noOptionsMessage": [Function],
                                "onChange": [Function],
                                "onInputChange": [Function],
                                "onMenuClose": [Function],
                                "onMenuOpen": [Function],
                                "openMenuOnClick": true,
                                "openMenuOnFocus": false,
                                "options": Array [],
                                "pageSize": 5,
                                "placeholder": "Select...",
                                "screenReaderStatus": [Function],
                                "styles": Object {},
                                "tabIndex": "0",
                                "tabSelectsValue": true,
                                "value": null,
                              }
                            }
                            setValue={[Function]}
                          >
                            <div
                              className="css-1ep9fjw "
                              onMouseDown={[Function]}
                              onTouchEnd={[Function]}
                              role="button"
                            >
                              <DownChevron>
                                <Svg
                                  focusable="false"
                                  role="presentation"
                                  size={20}
                                >
                                  <svg
                                    className="css-19bqh2r"
                                    focusable="false"
                                    height={20}
                                    role="presentation"
                                    viewBox="0 0 20 20"
                                    width={20}
                                  >
                                    <path
                                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                                    />
                                  </svg>
                                </Svg>
                              </DownChevron>
                            </div>
                          </DropdownIndicator>
                        </div>
                      </Component>
                    </IndicatorsContainer>
                  </div>
                </Component>
              </Control>
            </div>
          </Component>
        </SelectContainer>
      </Select>
    </Creatable>
  </StateManager>
</Async>
`;
